examples/jsm/objects/Fire.d.ts

```diff
@@ -34,7 +34,7 @@ export interface FireOptions {
 	massConservation?: boolean;
 }
 
-export class Fire extends Mesh {
+export class Fire extends Mesh<ShaderMaterial> {
```

Comment:
There are some more TS files that define classes which extend from `Mesh` (e.g. `Lensflare`). Is it necessary to change these definitions in the same way? I mean adding `<some material>`.

Reply:
In case of [Lensflare](https://github.com/mrdoob/three.js/blob/e8145dcaa83f57da82a3d3d413c89e34b9f01571/examples/jsm/objects/Lensflare.d.ts) it isn't because it does not override the `material` property with some different type.

In other cases, where some type extends `Mesh` and overrides the `material` property, it is necessary to change the definition, because doing so ensures exact typings all the way through.

An alternate approach would be to let the generic `TMaterial` of the `Mesh` class default to `Material` (which seems to be the base of all materials) then it shouldn't be necessary to change `Fire.d.ts`, but with the consequences of the `material` property in `Fire` class not being exact typed.

Reply:
# The Alternate approach I mentioned earlier

> An alternate approach would be to let the generic TMaterial of the Mesh class default to 
> Material then it shouldn't be necessary to change Fire.d.ts, but with the consequences of the material property in Fire class not being exactly typed.

Using this approach shouldn't break any existing code by making some compromises.

So, let the generic `TMaterial` of the `Mesh` class default to `Material`

**Mesh.d.ts**
```diff
- export class Mesh<TMaterial extends Material | Material[] = MeshBasicMaterial> extends Object3D {
+ export class Mesh<TMaterial extends Material | Material[] = Material> extends Object3D {
```
**Filre.d.ts** stays as it allways was.

## Using it
```js
new Mesh(geo).material // -> shows type Material (in reality it is MeshBasicMaterial)
// ... all other behave like before:
new Mesh(geo, new MeshBasicMaterial(...)).material // -> shows type MeshBasicMaterial
new Mesh(geo, [new MeshBasicMaterial(...)]).material // -> shows type MeshBasicMaterial[]

const mesh = new Mesh(geo, [
   new MeshBasicMaterial(), 
   new MeshNormalMaterial(), 
   new MeshLambertMaterial(),
   new MeshNormalMaterial(),
]);
mesh.material; // shows -> (MeshBasicMaterial | MeshNormalMaterial | MeshLambertMaterial)[]
```

## Extending it
### case 1:

```js
class Lensflare extends Mesh {
   // the property `material` stays unchanged.
}
new Lensflare().material // shows -> `Material` (in reality it is MeshBasicMaterial)
```

### case 2:
```js
class Fire extends Mesh {
// override the material prop with different type:
   material: ShaderMaterial;
}
new Fire().material 
// should "normally" show -> `Material` (in reality it is ShaderMaterial)
// note: vscode is able to detect the real type here and shows `ShaderMaterial`
```

### case 3:
```js
// adding the type as generic argument on the base class
class FireWithGenericMeshClass extends Mesh<ShaderMaterial> {
   material: ShaderMaterial;
}
new FireWithGenericMeshClass().material // shows -> ShaderMaterial
```

### case 4:
```js
class Fire1 extends Mesh {
    constructor(/*...*/){
        super(/*...*/);
        
        // set the material prop to some different type:
        this.material = new ShaderMaterial();
    }
 }
 new Fire1().material // shows -> `Material` (in reality it is ShaderMaterial)
```

## Type checking
```js
class Fire2 extends Mesh<ShaderMaterial> {
 }
 new Fire2().material // shows -> ShaderMaterial

 // assiging the wrong type shows some type error!
 new Fire2().material = new MeshNormalMaterial(); // ❌
```

```js
var mesh = new Mesh(geo, [new MeshNormalMaterial(), new ShaderMaterial()]);
// assiging the wrong type shows some type error!
mesh.material = new MeshNormalMaterial(); // ❌
```

```js
{
    const myMaterial = [new MeshNormalMaterial(), new ShaderMaterial()];
    const mesh = new Mesh(geo, myMaterial);
    // pushing a wrong type to the array shows some type error!
    mesh.material.push(new MeshBasicMaterial()); // ❌
}
{
    // the type error can be fixed by specifing a common base type.
    const myMaterial: Material[] = [new MeshNormalMaterial(), new ShaderMaterial()];
    const mesh = new Mesh(geo, myMaterial);
    mesh.material.push(new MeshBasicMaterial()); // ✔
}
```